hugr-model/src/v0/mod.rs

```diff
@@ -0,0 +1,368 @@
+    pub struct CallFunc {
+        /// The name of the function to be called.
+        pub name: Symbol,
+    }
+
+    /// See [`Operation::LoadFunc`].
+    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
+    pub struct LoadFunc {
+        /// The name of the function to be loaded.
+        pub name: Symbol,
+    }
+
+    /// See [`Operation::Custom`].
+    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
+    pub struct Custom {
+        /// The name of the custom operation.
+        pub name: Symbol,
+    }
+
+    /// See [`Operation::DefineAlias`].
+    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
+    pub struct DefineAlias {
+        /// The name of the alias to be defined.
+        pub name: Symbol,
+        /// The value of the alias.
+        pub value: TermId,
+    }
+
+    /// See [`Operation::DeclareAlias`].
+    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
+    pub struct DeclareAlias {
+        /// The name of the alias to be declared.
+        pub name: Symbol,
+        /// The type of the alias.
+        pub r#type: TermId,
+    }
+}
+
+/// A metadata item.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct MetaItem {
+    /// Name of the metadata item.
+    pub name: SmolStr,
+    /// Value of the metadata item.
+    pub value: Term,
+}
+
+/// A port in the graph.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Port {
+    /// The id of the edge that the port is connected to.
+    /// All ports referencing the same edge are connected.
+    pub edge: EdgeId,
+
+    /// The type of the port.
+    ///
+    /// This must be a term of type `Type`.
+    /// If the type is unknown, this will be a wildcard term.
+    pub r#type: TermId,
+
+    /// Metadata attached to the port.
+    pub meta: Vec<MetaItem>,
+}
+
+/// Schemes are parameterized terms.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Scheme {
+    /// The named parameters of the scheme.
+    ///
+    /// Within any term, the previous terms of the parameter list are available as variables.
+    pub params: Vec<SchemeParam>,
+    /// Constraints on the parameters of the scheme.
+    ///
+    /// All parameters are available as variables within the constraints.
+    /// The constraints must be terms of type `Constraint`.
+    pub constraints: TinyVec<[TermId; 3]>,
+    /// The body of the scheme.
+    ///
+    /// All parameters are available as variables within the body.
+    pub body: TermId,
+}
+
+/// A named parameter of a scheme.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct SchemeParam {
+    /// The name of the parameter.
+    pub name: TermVar,
+    /// The type of the parameter.
+    pub r#type: TermId,
+}
+
+/// A term in the compile time meta language.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum Term {
+    /// Standin for any term.
+    Wildcard,
+    /// The type of types.
+    Type,
+    /// The type of constraints.
+    Constraint,
+    /// A variable.
+    Var(TermVar),
+    /// A symbolic function application.
```

Comment:
Whoa....what's this? We are not doing dependent types, right?

Reply:
If the LHS is always a type scheme, and the right are TypeArg's (i.e. this is the equivalent of `ListType` where the LHS comes from an extension), then I think we should make that clear; I imagine most users will be familiar with generics but may not be coming from a type-theoretic background where they see "function application" and `MyList<T>` are the same thing...

Reply:
A "function application" applies a function on the LHS to some argument(s) on the RHS, yes?